### PR TITLE
JVM_IR: add $assertionsDisabled when an inlined function uses it

### PR DESCRIPTION
**Pull request by pyos**
NOTE: jvmCrossinlineLambdaDeclarationSite.kt is muted because the inliner does not remap references to an anonymous object's parent class after regenerating it. Unlike the JVM backend, JVM_IR uses the top level named class' assertion status for all inner classes. (The test used to pass because the lambda in `inline fun call` read the `$assertionsDisabled` field of `CrossinlineLambdaContainer`, which was not reloaded after changing the assertion status of package `test`.)

### CODE REVIEW

**Comment by ilmirus** [comment]
I have some concerns about the change. Let's consider, for example, one of your tests:
```
// FILE: InlineMe.kt
inline fun inlineMe(crossinline c : () -> String) = {
    assert(true)
    c()
}
// FILE: box.kt
fun box(): String = (inlineMe { "OK" })()
```
Let's compile the files: `$dist/kotlinc/bin/kotlinc -Xuse-ir -Xassertions=jvm InlineMe.kt box.kt`
The resulting bytecode has no mention of `$assertionsDisabled`.
This is not what I expect. I expect at least `BoxKt.class` to have assertion field generated, ideally, both `BoxKt` and `InlineMeKt` to have one. We need to generate the field in `InlineMeKt`, since one can call inline functions from Java or via reflection.
However, both `BoxKt` and `InlineMeKt` lack the field, rendering the assert statement useless.
Even replacing `assert(true)` with `assert(check())`, where `check` is `fun check() = true` to prevent unwanted constant folding, did not help.

File: compiler/testData/codegen/boxInline/assert/jvmAssertInlineFunctionAssertionsEnabled.kt [hunk 1]
```diff
@@ -28,6 +26,7 @@ class Dummy
 fun enableAssertions(): CheckerJvmAssertInlineFunctionAssertionsEnabled {
     val loader = Dummy::class.java.classLoader
     loader.setClassAssertionStatus("CheckerJvmAssertInlineFunctionAssertionsEnabled", true)
+    loader.setClassAssertionStatus("InlineKt", false)
```
[the remark this answers is not in the document]
Comment (pyos) [reply compiler/testData/codegen/boxInline/assert/jvmAssertInlineFunctionAssertionsEnabled.kt]: Technically yes, but IMO this makes the expected behavior more clear.

**Comment by pyos** [comment]
>However, both BoxKt and InlineMeKt lack the field, rendering the assert statement useless.

Absence of a field would not make the assertion useless, it would make it throw a NoSuchFieldException.

In any case, note the `=` in the definition of `inlineMe`: the field is generated in the callable reference object (`InlineMeKt$inlineMe$1` and `BoxKt$box$$inlined$inlineMe$1`), not in the file class. This is precisely the point of moving the lowerings: if the field is placed in `InlineMeKt` instead, the inliner would still remap the access to the regenerated class, but the copy will not have the field, causing the aforementioned NoSuchFieldException.

**Comment by ilmirus** [comment]
My bad. I got confused by "top named class" and though, that you refer to generated classfile.

After changing the test to
```
class A {
    inline fun inlineMe(crossinline c : () -> String) = {
        assert(true)
        c()
    }
}

class B {
    fun box(): String = (A().inlineMe { "OK" })()
}
```
`A.class` clearly contains the field. The inliner should put the field to `B.class`, but this is a bug in the inliner (and I am responsible for the bug).

Could you, please, add a test like
```
// FILE: A.kt

class A {
    inline fun inlineMe(crossinline c : () -> String) = {
        assert(true)
        c()
    }
}

// FILE: B.java

public class B {
	public static String check() {
		return new A().inlineMe(new Function0<String> {
			@Override String invoke() { return "OK" };
		}).invoke();
	}
}

// FILE: box.kt

fun box(): String = B.check()
```
?

Otherwise, LGTM.

**Comment by pyos** [comment]
Added the test.

**Comment by ilmirus** [comment]
Thanks!

**Pull request merged**